lib/src/annotate.rs

```diff
@@ -0,0 +1,394 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Methods that allow annotation (attribution and blame) for a file in a
+//! repository.
+//!
+//! TODO: Add support for different blame layers with a trait in the future.
+//! Like commit metadata and more.
+
+use std::collections::HashMap;
+
+use pollster::FutureExt;
+
+use crate::backend::BackendError;
+use crate::backend::CommitId;
+use crate::commit::Commit;
+use crate::conflicts::materialize_merge_result;
+use crate::conflicts::materialize_tree_value;
+use crate::conflicts::MaterializedTreeValue;
+use crate::diff::Diff;
+use crate::diff::DiffHunk;
+use crate::fileset::FilesetExpression;
+use crate::graph::GraphEdge;
+use crate::graph::GraphEdgeType;
+use crate::merged_tree::MergedTree;
+use crate::repo::Repo;
+use crate::repo_path::RepoPath;
+use crate::revset::RevsetEvaluationError;
+use crate::revset::RevsetExpression;
+use crate::revset::RevsetFilterPredicate;
+use crate::store::Store;
+
+/// Annotation results for a specific file
+pub struct AnnotateResults {
+    /// An array of annotation results ordered by line.
+    /// For each value in the array, the commit_id is the commit id of the
+    /// originator of the line and the string is the actual line itself (without
+    /// newline terminators). The vector is ordered by appearance in the
+    /// file
+    pub file_annotations: Vec<(CommitId, Vec<u8>)>,
+}
+
+/// A map from commits to line mappings.
+/// Namely, for a given commit A, the value is the mapping of lines in the file
+/// at commit A to line numbers in the original file
+type CommitLineMap = HashMap<CommitId, HashMap<usize, usize>>;
+
+/// Memoizes the file contents for a given version to save time
+type FileCache = HashMap<CommitId, Vec<u8>>;
```

Comment:
It might be simpler to combine `CommitLineMap` and `FileCache`. They are conceptually separate, but their lifetimes are the same?

Doing that will probably help remove `.unwrap()` from `process_files_in_commits()`. If the cache entry of the current commit has been removed from the `HashMap`, new entry can be inserted to the map without dropping the previous borrow.